services/rsources/handler.go

```
@@ -103,78 +104,128 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	defer func() { _ = stmt.Close() }()
 
 	for i := range records {
-		_, err = stmt.ExecContext(
+		if _, err = stmt.ExecContext(
 			ctx,
-			jobRunId,
-			key.TaskRunID,
-			key.SourceID,
-			key.DestinationID,
-			records[i])
-		if err != nil {
-			return err
+			id,
+			records[i]); err != nil {
+			return fmt.Errorf("inserting into rsources_failed_keys_v2_records: %w", err)
 		}
 	}
-
-	return
+	return nil
 }
 
-func (sh *sourcesHandler) GetFailedRecords(ctx context.Context, jobRunId string, filter JobFilter) (JobFailedRecords, error) {
+func (sh *sourcesHandler) GetFailedRecords(ctx context.Context, jobRunId string, filter JobFilter, paging PagingInfo) (JobFailedRecords, error) {
 	if sh.config.SkipFailedRecordsCollection {
 		return JobFailedRecords{ID: jobRunId}, ErrOperationNotSupported
 	}
-	filters, filterParams := sqlFilters(jobRunId, filter)
+	// first find the list of ids (postgres query planner uses an inefficient plan if there is one id with millions of records and a few ids with a few records)
+	ids, err := func() ([]string, error) {
+		var ids []string
+		filters, params := sqlFilters(jobRunId, filter)
+		rows, err := sh.readDB().QueryContext(ctx, `SELECT id FROM "rsources_failed_keys_v2" `+filters, params...)
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = rows.Close() }()
+		for rows.Next() {
+			var id string
+			if err := rows.Scan(&id); err != nil {
+				return nil, err
+			}
+			ids = append(ids, id)
+		}
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		return ids, nil
+	}()
+	if err != nil {
+		return JobFailedRecords{ID: jobRunId}, fmt.Errorf("failed to get failed record ids: %w", err)
+	}
 
+	filters := "WHERE r.id = ANY($1)"
+	params := []interface{}{pq.Array(ids)}
+	var limit string
+	if paging.Size > 0 {
+		if nextPageToken, err := NextPageTokenFromString(paging.NextPageToken); err == nil {
+			filters = filters + fmt.Sprintf(` AND r.id >= $%[1]d AND r.record_id > $%[2]d`, len(params)+1, len(params)+2)
```

If this unmarshal fails, we'd end up returning all the rows defeating the purpose of pagination.
We can return an error so the user knows to send the proper params.